lib/gringotts/gateways/mercadopago.ex

```diff
@@ -0,0 +1,308 @@
+    else
+      {state, res} = get_customer_id(opts)
+      if state == :error do
+        {state, res}
+      else
+        auth_token(amount, card, opts, res, false)
+      end
+    end
+  end
+  
+
+  @doc """
+  Captures a pre-authorized `amount`.
+
+  `amount` is transferred to the merchant account by mercadopago used in the
+  pre-authorization referenced by `payment_id`.
+
+  ## Note
+  mercadopago allows partial captures also. However, you can make a partial capture to a payment only **once**.
+
+  > The authorization will be valid for 7 days. If you do not capture it by that time, it will be cancelled.
+
+  > The specified amount can not exceed the originally reserved.
+
+  > If you do not specify the amount, all the reserved money is captured.
+  
+  > In Argentina only available for Visa and American Express cards.
+
+  ## Example
+
+  The following example shows how one would (partially) capture a previously
+  authorized a payment worth 35 BRL by referencing the obtained authorization `id`.
+
+      iex> amount = Money.new(35, :BRL)
+      iex> {:ok, capture_result} = Gringotts.capture(Gringotts.Gateways.Mercadopago, auth_result.id, amount, opts)
+  """
+  @spec capture(String.t(), Money.t, keyword) :: {:ok | :error, Response}
+  def capture(payment_id, amount, opts) do
+    {_, value, _, _} = Money.to_integer_exp(amount)
+    url = "#{@base_url}/v1/payments/#{payment_id}?access_token=#{opts[:config][:access_token]}"
+    body = %{"capture": true, "transaction_amount": value} |> Poison.encode!
+    headers = [{"content-type", "application/json"}, {"accept", "application/json"}]
+    commit(:put, url, body, headers) |> respond(opts)
+  end
+
+  
+
+
+
+
+
+  ###############################################################################
+  #                                PRIVATE METHODS                              #
+  ###############################################################################
+  
+  # Makes the request to mercadopago's network.
+  # For consistency with other gateway implementations, make your (final)
+  # network request in here, and parse it using another private method called
+  # `respond`.
+  #@spec commit(_, _, _, _) :: {:ok | :error, Response}
+  defp commit(method, path, body, headers) do
+    HTTPoison.request(method, path, body, headers, [])
+  end
+
+  # Parses mercadopago's response and returns a `Gringotts.Response` struct
+  # in a `:ok`, `:error` tuple.
+
+  defp auth_token(amount, %CreditCard{} = card, opts, customer_id, capture) do
+    opts = opts ++ [customer_id: customer_id]
+    {state, res} = get_token_id(card, opts)
+    if state == :error do
+      {state, res}
+    else
+      auth(amount, card, opts, res, capture)
+    end
+  end
+
+  defp auth(amount, %CreditCard{} = card, opts, token_id, capture) do
+    {_, value, _, _} = Money.to_integer_exp(amount)
+    opts = opts ++ [token_id: token_id]
+    url = "#{@base_url}/v1/payments?access_token=#{opts[:config][:access_token]}"
```

Comment:
We cannot move `url` into commit, because it if different for different cases. We have moved the `headers` and `respond` function call in commit itself.

Reply:
You can (partially)! Notice that [`HTTPoison.request/5`](https://hexdocs.pm/httpoison/HTTPoison.html#request/5) takes many `options`, one of them being the key `:params`.
```elixir
auth(amount, %CreditCard{} = card, opts, token_id, capture) do
  {_, value, _, _} = Money.to_integer_exp(amount)
  opts = opts ++ [token_id: token_id]
  url = "/v1/payments"
  url_params = [access_token: opts[:config][:access_token]]
  body = authorize_params(value, card, opts, opts[:token_id], opts[:customer_id], capture) |> Poison.encode!
  commit(:post, url, body, params: url_params)
end

defp commit(method, path, body, opts) do
  headers = [{"content-type", "application/json"}, {"accept", "application/json"}]
  HTTPoison.request(method, path, body, headers, opts) |> respond()
end
```